File: src/domain/context-pack.ts
import type { BaseObject, IsoDateTime, StableId } from "./common";
import { createToyBaseObject } from "./common";

export const CONTEXT_TASK_TYPES = [
  "source-breakdown",
  "creative-discussion",
  "outline-generation",
  "character-simulation",
  "scene-planning",
  "draft-generation",
  "style-revision",
  "continuity-audit",
  "writeback-suggestion",
] as const;

export type ContextTaskType = (typeof CONTEXT_TASK_TYPES)[number];
export type SourceBreakdownStage =
  | "segment-extraction"
  | "parent-synthesis"
  | "cross-dimension-review"
  | "draft-review-rerun";
export interface ContextTaskDescriptor {
  taskType?: ContextTaskType;
  sourceBreakdownStage: SourceBreakdownStage;
  inputRefs?: StableId[];
  outputRefs?: StableId[];
  sourceChunkRefs?: StableId[];
  originalChunkRef?: StableId;
  originalDraftRef?: StableId;
  userFeedbackRef?: StableId;
  revisionRequestRef?: StableId;
  outputSchemaRef?: string;
  outputSchemaVersion?: number;
  contextHash?: string;
}
export type ActivationReason =
  | "target-match"
  | "direct-reference"
  | "context-hint-keyword"
  | "task-default"
  | "change-set-impact"
  | "continuity-dependency"
  | "user-selected"
  | "pinned-required";
export type EvidenceChainEntry = {
  ref: StableId;
  relation: string;
  note: string;
};
export type FreshnessState = "current" | "stale" | "superseded" | "unknown";
export type UserSelectionState = "auto" | "user-included" | "user-excluded" | "user-pinned";
export type ContextPriority = "required" | "high" | "normal" | "low";
export type RelevanceBand = "low" | "medium" | "high";
export type ConfidenceLevel = "low" | "medium" | "high";

export interface ContextActivation {
  id: StableId;
  contextPackId: StableId;
  sourceRef: StableId;
  sourceObjectType: string;
  sourceSchemaVersion: number;
  activationReason: ActivationReason;
  evidenceChain: EvidenceChainEntry[];
  relevance: RelevanceBand;
  confidence: ConfidenceLevel;
  priority: ContextPriority;
  freshnessState: FreshnessState;
  userSelection: UserSelectionState;
  selectedAt: IsoDateTime;
}

export interface TokenBudget {
  maxInputTokens: number;
  reservedOutputTokens: number;
  estimationMethod: string;
}

export interface FragmentBudget {
  maxFragments: number;
  maxFragmentChars: number;
}

export interface ObjectBudget {
  objectType: string;
  maxObjects: number;
}

export interface PriorityBands {
  required: number;
  high: number;
  normal: number;
  low: number;
}

export type TrimmingStrategy =
  | "drop-low-priority"
  | "summarize-eligible"
  | "latest-confirmed-first"
  | "closest-scope-first"
  | "user-pinned-first";
export type OverflowPolicy = "fail-closed" | "drop-with-exclusion-record";

export interface ContextBudget {
  id: StableId;
  contextPackId: StableId;
  tokenBudget: TokenBudget;
  fragmentBudget: FragmentBudget;
  objectBudget: ObjectBudget[];
  priorityBands: PriorityBands;
  trimmingStrategy: TrimmingStrategy;
  excludedObjectTypes: string[];
  excludedRefs: StableId[];
  overflowPolicy: OverflowPolicy;
}

export interface ManifestHashEntry {
  ref: StableId;
  hash: string;
}

export interface SchemaVersionEntry {
  schemaId: string;
  version: number;
}

export interface ContextHash {
  algorithm: "sha256";
  canonicalization: "json-canonical-v1";
  schemaVersionMap: SchemaVersionEntry[];
  libraryManifestHash: string;
  sourceHashRefs: ManifestHashEntry[];
  changeSetRefs: StableId[];
  changeSetHash: string;
  selectionPolicyVersion: string;
  promptTemplateVersion?: string;
  includedItemHashes: ManifestHashEntry[];
  activationHashes: ManifestHashEntry[];
  excludedItemHashes?: ManifestHashEntry[];
  value: string;
}

export type ContextExclusionReason =
  | "budget-insufficient"
  | "stale-or-superseded"
  | "user-excluded"
  | "permission-boundary"
  | "task-irrelevant"
  | "insufficient-evidence"
  | "duplicate-or-covered"
  | "schema-unsupported"
  | "source-boundary"
  | "not-in-selected-scope";

export type ContextExclusionRiskKind =
  | "privacy-boundary"
  | "copyright-boundary"
  | "stale-state"
  | "budget-overflow"
  | "irrelevant"
  | "low-confidence"
  | "user-choice"
  | "permission-denied"
  | "unsupported-schema";

export interface ContextExclusion {
  id: StableId;
  contextPackId: StableId;
  excludedRef: StableId;
  sourceObjectType: string;
  reason: ContextExclusionReason;
  riskKind: ContextExclusionRiskKind;
  explanation: string;
  candidateScore?: RelevanceBand;
  wouldFitIfBudget?: boolean;
  relatedActivationRef?: StableId;
  createdAt: IsoDateTime;
}

export interface ContextIncludedItem {
  sourceRef: StableId;
  sourceObjectType: string;
  sourceSchemaVersion: number;
  contentHash: string;
  activationRef: StableId;
}

export interface ContextPack extends BaseObject {
  type: "context-pack";
  taskType: ContextTaskType;
  taskDescriptor?: ContextTaskDescriptor;
  targetRef: StableId;
  activations: ContextActivation[];
  includedItems: ContextIncludedItem[];
  exclusions: ContextExclusion[];
  budget: ContextBudget;
  contextHash: ContextHash;
  createdForTaskRef: StableId;
  createdForJobRef?: StableId;
}

const HASH_A = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
const HASH_B = "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb";
const HASH_C = "cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc";
const HASH_D = "dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd";
const HASH_E = "eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee";
const HASH_F = "ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff";

export function createToyDraftGenerationContextPack(timestamp: IsoDateTime): ContextPack {
  const base = createToyBaseObject(
    "context-pack",
    "context-pack-draft-toy",
    "Toy Draft Generation Context",
    timestamp,
    "draft",
  );

  const activations: ContextActivation[] = [
    createActivation(
      "activation-scene-toy",
      "scene-card-toy",
      "SceneCard",
      "target-match",
      "required",
      "current",
      timestamp,
    ),
    createActivation(
      "activation-beat-toy",
      "beat-toy",
      "Beat",
      "direct-reference",
      "high",
      "current",
      timestamp,
    ),
    createActivation(
      "activation-character-state-toy",
      "character-state-toy",
      "CharacterDynamicState",
      "task-default",
      "high",
      "current",
      timestamp,
    ),
  ];

  return {
    ...base,
    taskType: "draft-generation",
    targetRef: "scene-card-toy",
    activations,
    includedItems: [
      includedItem("scene-card-toy", "SceneCard", HASH_A, "activation-scene-toy"),
      includedItem("beat-toy", "Beat", HASH_B, "activation-beat-toy"),
      includedItem(
        "character-state-toy",
        "CharacterDynamicState",
        HASH_C,
        "activation-character-state-toy",
      ),
    ],
    exclusions: [
      {
        id: "exclusion-style-budget-toy",
        contextPackId: "context-pack-draft-toy",
        excludedRef: "style-profile-extra-toy",
        sourceObjectType: "StyleProfile",
        reason: "budget-insufficient",
        riskKind: "budget-overflow",
        explanation: "Optional style profile exceeded the object budget for draft preview.",
        candidateScore: "medium",
        wouldFitIfBudget: true,
        createdAt: timestamp,
      },
      {
        id: "exclusion-state-stale-toy",
        contextPackId: "context-pack-draft-toy",
        excludedRef: "character-state-stale-toy",
        sourceObjectType: "CharacterDynamicState",
        reason: "stale-or-superseded",
        riskKind: "stale-state",
        explanation: "A confirmed change set superseded this older character state.",
        candidateScore: "high",
        wouldFitIfBudget: true,
        createdAt: timestamp,
      },
    ],
    budget: {
      id: "budget-draft-toy",
      contextPackId: "context-pack-draft-toy",
      tokenBudget: {
        maxInputTokens: 1200,
        reservedOutputTokens: 600,
        estimationMethod: "toy-char-estimate",
      },
      fragmentBudget: {
        maxFragments: 6,
        maxFragmentChars: 800,
      },
      objectBudget: [
        { objectType: "SceneCard", maxObjects: 1 },
        { objectType: "Beat", maxObjects: 4 },
        { objectType: "CharacterDynamicState", maxObjects: 1 },
      ],
      priorityBands: {
        required: 1,
        high: 3,
        normal: 2,
        low: 0,
      },
      trimmingStrategy: "drop-low-priority",
      excludedObjectTypes: [],
      excludedRefs: [],
      overflowPolicy: "fail-closed",
    },
    contextHash: {
      algorithm: "sha256",
      canonicalization: "json-canonical-v1",
      schemaVersionMap: [
        {
          schemaId: "https://infinite-typewriter.local/schemas/context-pack.schema.json",
          version: 1,
        },
      ],
      libraryManifestHash: HASH_D,
      sourceHashRefs: [{ ref: "source-toy", hash: HASH_E }],
      changeSetRefs: ["change-set-toy"],
      changeSetHash: HASH_F,
      selectionPolicyVersion: "context-pack-selection-v1",
      includedItemHashes: [
        { ref: "scene-card-toy", hash: HASH_A },
        { ref: "beat-toy", hash: HASH_B },
        { ref: "character-state-toy", hash: HASH_C },
      ],
      activationHashes: [
        { ref: "scene-card-toy", hash: HASH_A },
        { ref: "beat-toy", hash: HASH_B },
        { ref: "character-state-toy", hash: HASH_C },
      ],
      excludedItemHashes: [{ ref: "exclusion-manifest-toy", hash: HASH_D }],
      value: "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef",
    },
    createdForTaskRef: "task-draft-preview-toy",
  };
}

export function createToySourceBreakdownRevisionContextPack(timestamp: IsoDateTime): ContextPack {
  const base = createToyBaseObject(
    "context-pack",
    "context-pack-source-breakdown-rerun-toy",
    "Toy Source Breakdown Revision Context",
    timestamp,
    "draft",
  );
  const contextHash = "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef";

  return {
    ...base,
    taskType: "source-breakdown",
    taskDescriptor: {
      taskType: "source-breakdown",
      sourceBreakdownStage: "draft-review-rerun",
      inputRefs: ["source-chunk-toy", "draft-output-fake-toy", "user-feedback-toy"],
      outputRefs: ["revision-suggestion-toy"],
      sourceChunkRefs: ["source-chunk-toy"],
      originalChunkRef: "source-chunk-toy",
      originalDraftRef: "draft-output-fake-toy",
      userFeedbackRef: "user-feedback-toy",
      revisionRequestRef: "revision-request-toy",
      outputSchemaRef: "toy-draft-output",
      outputSchemaVersion: 1,
      contextHash,
    },
    targetRef: "revision-request-toy",
    activations: [
      createActivation(
        "activation-revision-request-toy",
        "revision-request-toy",
        "RevisionRequest",
        "target-match",
        "required",
        "current",
        timestamp,
        "context-pack-source-breakdown-rerun-toy",
      ),
    ],
    includedItems: [
      includedItem("source-chunk-toy", "SourceChunk", HASH_A, "activation-revision-request-toy"),
      includedItem("draft-output-fake-toy", "DraftOutput", HASH_B, "activation-revision-request-toy"),
      includedItem("user-feedback-toy", "UserFeedback", HASH_C, "activation-revision-request-toy"),
    ],
    exclusions: [],
    budget: {
      id: "budget-source-breakdown-rerun-toy",
      contextPackId: "context-pack-source-breakdown-rerun-toy",
      tokenBudget: {
        maxInputTokens: 900,
        reservedOutputTokens: 300,
        estimationMethod: "toy-char-estimate",
      },
      fragmentBudget: {
        maxFragments: 4,
        maxFragmentChars: 700,
      },
      objectBudget: [
        { objectType: "RevisionRequest", maxObjects: 1 },
        { objectType: "SourceChunk", maxObjects: 1 },
        { objectType: "DraftOutput", maxObjects: 1 },
        { objectType: "UserFeedback", maxObjects: 1 },
      ],
      priorityBands: {
        required: 1,
        high: 3,
        normal: 0,
        low: 0,
      },
      trimmingStrategy: "drop-low-priority",
      excludedObjectTypes: [],
      excludedRefs: [],
      overflowPolicy: "fail-closed",
    },
    contextHash: {
      algorithm: "sha256",
      canonicalization: "json-canonical-v1",
      schemaVersionMap: [
        {
          schemaId: "https://infinite-typewriter.local/schemas/context-pack.schema.json",
          version: 1,
        },
      ],
      libraryManifestHash: HASH_D,
      sourceHashRefs: [{ ref: "source-toy", hash: HASH_E }],
      changeSetRefs: ["change-set-toy"],
      changeSetHash: HASH_F,
      selectionPolicyVersion: "context-pack-selection-v1",
      promptTemplateVersion: "source-breakdown-rerun-v1",
      includedItemHashes: [
        { ref: "source-chunk-toy", hash: HASH_A },
        { ref: "draft-output-fake-toy", hash: HASH_B },
        { ref: "user-feedback-toy", hash: HASH_C },
      ],
      activationHashes: [{ ref: "revision-request-toy", hash: HASH_A }],
      value: contextHash,
    },
    createdForTaskRef: "task-source-breakdown-rerun-toy",
  };
}

function createActivation(
  id: StableId,
  sourceRef: StableId,
  sourceObjectType: string,
  activationReason: ActivationReason,
  priority: ContextPriority,
  freshnessState: FreshnessState,
  timestamp: IsoDateTime,
  contextPackId = "context-pack-draft-toy",
): ContextActivation {
  return {
    id,
    contextPackId,
    sourceRef,
    sourceObjectType,
    sourceSchemaVersion: 1,
    activationReason,
    evidenceChain: [
      {
        ref: "scene-card-toy",
        relation: "draft-context",
        note: "Toy public fixture links the target scene to required draft context.",
      },
    ],
    relevance: priority === "required" ? "high" : "medium",
    confidence: "high",
    priority,
    freshnessState,
    userSelection: "auto",
    selectedAt: timestamp,
  };
}

function includedItem(
  sourceRef: StableId,
  sourceObjectType: string,
  contentHash: string,
  activationRef: StableId,
): ContextIncludedItem {
  return {
    sourceRef,
    sourceObjectType,
    sourceSchemaVersion: 1,
    contentHash,
    activationRef,
  };
}

File: crates/core/src/model/context_pack.rs
use serde::{Deserialize, Serialize};

use super::common::{BaseObject, IsoDateTime, StableId};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ContextTaskType {
    SourceBreakdown,
    CreativeDiscussion,
    OutlineGeneration,
    CharacterSimulation,
    ScenePlanning,
    DraftGeneration,
    StyleRevision,
    ContinuityAudit,
    WritebackSuggestion,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum SourceBreakdownStage {
    SegmentExtraction,
    ParentSynthesis,
    CrossDimensionReview,
    DraftReviewRerun,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextTaskDescriptor {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub task_type: Option<ContextTaskType>,
    pub source_breakdown_stage: SourceBreakdownStage,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_refs: Option<Vec<StableId>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_refs: Option<Vec<StableId>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_chunk_refs: Option<Vec<StableId>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub original_chunk_ref: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub original_draft_ref: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_feedback_ref: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub revision_request_ref: Option<StableId>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_schema_ref: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_schema_version: Option<u16>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_hash: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ActivationReason {
    TargetMatch,
    DirectReference,
    ContextHintKeyword,
    TaskDefault,
    ChangeSetImpact,
    ContinuityDependency,
    UserSelected,
    PinnedRequired,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceChainEntry {
    pub r#ref: StableId,
    pub relation: String,
    pub note: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum RelevanceBand {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ConfidenceLevel {
    Low,
    Medium,
    High,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ContextPriority {
    Required,
    High,
    Normal,
    Low,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum FreshnessState {
    Current,
    Stale,
    Superseded,
    Unknown,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum UserSelectionState {
    Auto,
    UserIncluded,
    UserExcluded,
    UserPinned,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextActivation {
    pub id: StableId,
    pub context_pack_id: StableId,
    pub source_ref: StableId,
    pub source_object_type: String,
    pub source_schema_version: u16,
    pub activation_reason: ActivationReason,
    pub evidence_chain: Vec<EvidenceChainEntry>,
    pub relevance: RelevanceBand,
    pub confidence: ConfidenceLevel,
    pub priority: ContextPriority,
    pub freshness_state: FreshnessState,
    pub user_selection: UserSelectionState,
    pub selected_at: IsoDateTime,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenBudget {
    pub max_input_tokens: u32,
    pub reserved_output_tokens: u32,
    pub estimation_method: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FragmentBudget {
    pub max_fragments: u32,
    pub max_fragment_chars: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ObjectBudget {
    pub object_type: String,
    pub max_objects: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PriorityBands {
    pub required: u32,
    pub high: u32,
    pub normal: u32,
    pub low: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum TrimmingStrategy {
    DropLowPriority,
    SummarizeEligible,
    LatestConfirmedFirst,
    ClosestScopeFirst,
    UserPinnedFirst,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum OverflowPolicy {
    FailClosed,
    DropWithExclusionRecord,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextBudget {
    pub id: StableId,
    pub context_pack_id: StableId,
    pub token_budget: TokenBudget,
    pub fragment_budget: FragmentBudget,
    pub object_budget: Vec<ObjectBudget>,
    pub priority_bands: PriorityBands,
    pub trimming_strategy: TrimmingStrategy,
    pub excluded_object_types: Vec<String>,
    pub excluded_refs: Vec<StableId>,
    pub overflow_policy: OverflowPolicy,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ManifestHashEntry {
    pub r#ref: StableId,
    pub hash: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaVersionEntry {
    pub schema_id: String,
    pub version: u16,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum HashAlgorithm {
    Sha256,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Canonicalization {
    JsonCanonicalV1,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextHash {
    pub algorithm: HashAlgorithm,
    pub canonicalization: Canonicalization,
    pub schema_version_map: Vec<SchemaVersionEntry>,
    pub library_manifest_hash: String,
    pub source_hash_refs: Vec<ManifestHashEntry>,
    pub change_set_refs: Vec<StableId>,
    pub change_set_hash: String,
    pub selection_policy_version: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub prompt_template_version: Option<String>,
    pub included_item_hashes: Vec<ManifestHashEntry>,
    pub activation_hashes: Vec<ManifestHashEntry>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub excluded_item_hashes: Option<Vec<ManifestHashEntry>>,
    pub value: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ContextExclusionReason {
    BudgetInsufficient,
    StaleOrSuperseded,
    UserExcluded,
    PermissionBoundary,
    TaskIrrelevant,
    InsufficientEvidence,
    DuplicateOrCovered,
    SchemaUnsupported,
    SourceBoundary,
    NotInSelectedScope,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum ContextExclusionRiskKind {
    PrivacyBoundary,
    CopyrightBoundary,
    StaleState,
    BudgetOverflow,
    Irrelevant,
    LowConfidence,
    UserChoice,
    PermissionDenied,
    UnsupportedSchema,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextExclusion {
    pub id: StableId,
    pub context_pack_id: StableId,
    pub excluded_ref: StableId,
    pub source_object_type: String,
    pub reason: ContextExclusionReason,
    pub risk_kind: ContextExclusionRiskKind,
    pub explanation: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub candidate_score: Option<RelevanceBand>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub would_fit_if_budget: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub related_activation_ref: Option<StableId>,
    pub created_at: IsoDateTime,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextIncludedItem {
    pub source_ref: StableId,
    pub source_object_type: String,
    pub source_schema_version: u16,
    pub content_hash: String,
    pub activation_ref: StableId,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContextPack {
    #[serde(flatten)]
    pub base: BaseObject,
    pub task_type: ContextTaskType,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub task_descriptor: Option<ContextTaskDescriptor>,
    pub target_ref: StableId,
    pub activations: Vec<ContextActivation>,
    pub included_items: Vec<ContextIncludedItem>,
    pub exclusions: Vec<ContextExclusion>,
    pub budget: ContextBudget,
    pub context_hash: ContextHash,
    pub created_for_task_ref: StableId,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub created_for_job_ref: Option<StableId>,
}
